pylint/checkers/variables.py

```diff
@@ -2220,24 +2220,17 @@ def _is_variable_violation(
                     )
                 ):
                     # Exempt those definitions that are used inside the type checking
-                    # guard or that are defined in both type checking guard branches.
+                    # guard or that are defined in any elif/else type checking guard branches.
                     used_in_branch = defstmt_parent.parent_of(node)
-                    defined_in_or_else = False
-
-                    for definition in defstmt_parent.orelse:
-                        if isinstance(definition, nodes.Assign):
-                            defined_in_or_else = any(
-                                target.name == node.name
-                                for target in definition.targets
-                                if isinstance(target, nodes.AssignName)
-                            )
-                        elif isinstance(
-                            definition, (nodes.ClassDef, nodes.FunctionDef)
-                        ):
-                            defined_in_or_else = definition.name == node.name
-
-                        if defined_in_or_else:
-                            break
+                    if defstmt_parent.has_elif_block():
+                        defined_in_or_else = utils.is_defined(
+                            node.name, defstmt_parent.orelse[0]
+                        )
+                    else:
+                        defined_in_or_else = any(
+                            utils.is_defined(node.name, content)
+                            for content in defstmt_parent.orelse
+                        )
 
                     if not used_in_branch and not defined_in_or_else:
                         maybe_before_assign = True
```

Comment:
It takes some effort to define `defined_in_or_else`, but it looks like the value won't get used if `used_in_branch` is true. Would something like this work?

```suggestion
                if not used_in_branch:
                    if defstmt_parent.has_elif_block():
                        defined_in_or_else = utils.is_defined(
                            node.name, defstmt_parent.orelse[0]
                        )
                    else:
                        defined_in_or_else = any(
                            utils.is_defined(node.name, content)
                            for content in defstmt_parent.orelse
                        )

                    if not defined_in_or_else:
                        maybe_before_assign = True
```

I think that should be equivalent, but the existing logic is very unpleasant and hard to understand, so please double check.

With a little further further shuffling, it might even be possible to get rid of a few more extra branches and eliminate the `too-many-branches` warning.

Reply:
Great suggestion! Refactored accordingly.

Regarding code readability, `is_variable_violation` is overloaded with logic and probably needs more function abstractions. But that is non-trivial and should probably be addressed in a separate PR.